target/scripts/startup/setup-stack.sh

```diff
@@ -1259,3 +1294,110 @@ function _setup_timezone
+      elif [[ -d ${FILE} ]]
+      then
+        _log 'trace' "Moving contents of ${FILE} to ${DEST}"
+        # Empty volume was mounted, or new content from enabling a feature ENV:
+        mv "${FILE}" "${DEST}"
+      fi
+
+      # Symlink the original path in the container ($FILE) to be
+      # sourced from assocaiated path in /var/mail-state/ ($DEST):
+      ln -s "${DEST}" "${FILE}"
+    done
+
+    # This ensures the user and group of the files from the external mount have their
+    # numeric ID values in sync. New releases where the installed packages order changes
+    # can change the values in the Docker image, causing an ownership mismatch.
+    # NOTE: More details about users and groups added during image builds are documented here:
+    # https://github.com/docker-mailserver/docker-mailserver/pull/3011#issuecomment-1399120252
+    _log 'trace' 'Fixing /var/mail-state/* permissions'
+    [[ ${ENABLE_AMAVIS}       -eq 1 ]] && chown -R amavis:amavis             /var/mail-state/lib-amavis
+    [[ ${ENABLE_CLAMAV}       -eq 1 ]] && chown -R clamav:clamav             /var/mail-state/lib-clamav
+    [[ ${ENABLE_FETCHMAIL}    -eq 1 ]] && chown -R fetchmail:nogroup         /var/mail-state/lib-fetchmail
+    [[ ${ENABLE_POSTGREY}     -eq 1 ]] && chown -R postgrey:postgrey         /var/mail-state/lib-postgrey
+    [[ ${ENABLE_RSPAMD}       -eq 1 ]] && chown -R _rspamd:_rspamd           /var/mail-state/lib-rspamd
+    [[ ${ENABLE_SPAMASSASSIN} -eq 1 ]] && chown -R debian-spamd:debian-spamd /var/mail-state/lib-spamassassin
+
+    chown -R root:root /var/mail-state/lib-logrotate
+    chown -R postfix:postfix /var/mail-state/lib-postfix
+
+    # NOTE: The Postfix spool location has mixed owner/groups to take into account:
+    # UID = postfix(101): active, bounce, corrupt, defer, deferred, flush, hold, incoming, maildrop, private, public, saved, trace
+    # UID = root(0): dev, etc, lib, pid, usr
+    # GID = postdrop(103): maildrop, public
+    # GID for all other directories is root(0)
+    # NOTE: `spool-postfix/private/` will be set to `postfix:postfix` when Postfix starts / restarts
+    # Set most common ownership:
+    chown -R postfix:root /var/mail-state/spool-postfix
+    chown root:root /var/mail-state/spool-postfix
+    # These two require the postdrop(103) group:
+    chgrp -R postdrop /var/mail-state/spool-postfix/maildrop
+    chgrp -R postdrop /var/mail-state/spool-postfix/public
+    # These all have root ownership at the src location:
+    chown -R root /var/mail-state/spool-postfix/dev
+    chown -R root /var/mail-state/spool-postfix/etc
+    chown -R root /var/mail-state/spool-postfix/lib
+    chown -R root /var/mail-state/spool-postfix/pid
+    chown -R root /var/mail-state/spool-postfix/usr
+  elif [[  ${ONE_DIR} -eq 1 ]]
+  then
+    _log 'warn' "ONE_DIR is enabled (=1), but '${STATEDIR}' does not exist - is a volume mounted there?"
+  else
+    _log 'debug' "Mail state is not consolidated into one directory"
+  fi
+}
```

Comment:
Suggested action: Extract into separate `setup.d` file, or keep `misc-stack.sh`

Technically fitting of `setup.d/one-dir.sh`, as it's an integration feature to support the `ONE_DIR` ENV paired with an external volume. As a purely integration focused feature though, it's a bit distinct from other `setup-stack.sh`, and is serving two functions (_`ONE_DIR=1` symlinks to `/var/mail-state`, and ensuring permissions/ownership is correct across restarts, notably during updating DMS releases when user and group IDs may have changed and need to adjust attributes of externally persisted state_).

---

The only change here from `misc-stack.sh` really is the extra conditional handling at the end to output a warning or debug log message.

I am a tad confused though. With the [commit expressing intent to split `setup-stack.sh` into smaller files for better scoping](https://github.com/docker-mailserver/docker-mailserver/pull/3112/commits/980d0a7b4662481cf51ed8835342abb1338cfdc4), you later [deleted `misc-stack.sh` to add it here in `setup-stack.sh`](https://github.com/docker-mailserver/docker-mailserver/pull/3112/commits/a65d9299bd451de4df8ebd5d05368fcf29a8a414), despite it being a good candidate as a separate file?

Personally, I'd envision this being refactored in future if we had better scoping of features. I preferred `misc-stack.sh` as a separate stage in it's current form, but would like services in setup to have a more clear cut scope where they'd have some convention to do their config setup, fix permissions/ownership like this method does, etc. At least for me it seems nicer if the bulk of Amavis + SA support and rspamd for example were mostly self-contained within their own folders.

Reply:
This will be moved in the follow-up after #3115 into a separate file in `setup.d`. I kept it in `setup.sh` because, yeah, TBH, I don't know either. Should have directly put it into `setup.d`.